Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,789 @@
+
+                switch action {
+                case .callDidYield:
+                    let delegate = self._delegate
+                    self._lock.unlock()
+                    delegate?.didYield(contentsOf: sequence)
+
+                case .throwError(let error):
+                    self._lock.unlock()
+                    throw error
+
+                case .suspendTask:
+                    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
+                        self._stateMachine.yield(
+                            contentsOf: sequence,
+                            continuation: continuation,
+                            yieldID: yieldID
+                        )
+
+                        self._lock.unlock()
+                    }
+                }
+            } onCancel: {
+                let action = self._lock.withLock { self._stateMachine.cancel(yieldID: yieldID) }
+
+                switch action {
+                case .resumeContinuationWithError(let continuation, let error):
+                    continuation.resume(throwing: error)
+
+                case .none:
+                    break
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func finish(with failure: Failure?) {
+            let delegate: Delegate? = self._lock.withLock {
+                let action = self._stateMachine.finish()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self._delegate
+                    self._delegate = nil
+
+                    return delegate
+
+                case .resumeContinuationsWithErrorAndCallDidTerminate(let suspendedYields, let error):
+                    let delegate = self._delegate
+                    self._delegate = nil
+                    // It is safe to resume the continuation while holding the lock
+                    // since the task will get enqueued on its executor and the resume method
+                    // is returning immediately
+                    suspendedYields.forEach { $0.continuation.resume(throwing: error) }
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate(failure: failure)
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    @usableFromInline
+    /* private */ internal struct StateMachine {
+        /// This is a small helper struct to encapsulate the three different values for a suspended yield.
+        @usableFromInline
+        /* private */ internal struct SuspendedYield {
+            /// The yield's ID.
+            @usableFromInline
+            var yieldID: UInt64
+            /// The yield's produced sequence of elements.
+            @usableFromInline
+            var elements: AnySequence<Element>
```

Comment:
Should the AsyncWriter be generic over the Sequence type as well? How often do user's want to send different Sequence types? For the SequenceOfOne/Two case users could always opt for an enum that implements Sequence.

Reply:
Yeah, I didn't like the point of holding an `AnySequence` here. If we do that, then we should get rid of the yielding of a single element.

Reply:
> If we do that, then we should get rid of the yielding of a single element.
I think we must and I consider this a good thing :P

Reply:
> Should the AsyncWriter be generic over the Sequence type as well? How often do user's want to send different Sequence types?

I'm not sure that's a good idea in practice. I think the typical usage pattern would be for a framework to create a writer and pass it to a caller. That would mean the framework would be responsible for choosing what the sequence type is _or_ the caller has to ask the framework for a writer with a given sequence type (which is a pretty awkward API).

> For the SequenceOfOne/Two case users could always opt for an enum that implements Sequence.

Not offering a yield-one API and requiring users to implement an enum like this makes for a really user hostile API IMHO.

---

FWIW I also don't like `AnySequence` here but I think we should keep it for the time being (it is an implementation detail and can be changed later). 

One alternative would be to always store `[Element]` and special case the API to avoid copies when an array is provided and copy into an array when anything else is provided. 

Taking that further we could be to pull a 'NIOAny' trick and have an enum of `array`, `one`, and `anySequence`.

That said, I think we should focus on getting the API right first and then compare perf.

Reply:
I agree with what George said. Making the whole thing generic of the sequence that is yielded severely limits the usage of the type and makes it very awkward to hold. I would stick with `AnySequence` for now and as George said we can always make it faster by applying some of our tricks.

Reply:
As discussed offline, I think we should change this to remove the `AnySequence`.